### DISCUSSION TITLE
VipsOperation: class "canny" not found

### DISCUSSION THREAD

**Discussion by Jacherr** [libvips/libvips#4048] [Q&A]
Attempting to call `vips_canny()` from C causes this, despite from CLI using `vips canny` works fine. I'm almost certain that they are using the same installation (8.15.2), because `vips --version` reports 8.15.2, `vips_version_string()` reports 8.15.2, and `pkg-config vips --modversion` also reports 8.15.2 - so no idea how this could be happening! Is it possible I'm missing some dependency?

**Comment by lovell** [comment]
The `vips_canny` operation was added about 6 years ago in v8.7.0 and has no external dependencies. I don't think it's possible to build libvips without this operation, unless you're patching or blocking it.

It sounds like there could be an old version of libvips being found at runtime. Are you able to use `ldd` with your compiled executable to see what shared libraries it's using?

**Comment by Jacherr** [comment]
I had a go at checking both my compiled executable and the vips executable, and they both point towards the same `/usr/local/lib64/libvips.so.42`. I've also searched `/usr` just to make sure there isn't another stray vips version somewhere else and there doesn't seem to be.
![image](https://github.com/user-attachments/assets/8a0d5524-8c40-4b01-82f1-6994f37344fe)

**Comment by jcupitt** [comment]
I don't know if this helps but I made a tiny test prog:

```C
/* compile with
 *      gcc -g -Wall canny.c `pkg-config vips --cflags --libs`
 */

#include <vips/vips.h>

int
main(int argc, char **argv)
{
    VipsImage *image;
    VipsImage *t1, *t2;

    if (VIPS_INIT(argv[0]))
        return -1;

    printf("built against libvips %s\n", VIPS_VERSION_STRING);
    printf("linked to libvips %s\n", vips_version_string());

    if (!(image = vips_image_new_from_file(argv[1], 
        "access", VIPS_ACCESS_SEQUENTIAL, 
        NULL)))
        vips_error_exit(NULL);
    
    if (vips_canny(image, &t1, "sigma", 1.5, NULL) ||
        vips_linear1(t1, &t2, 100, 0, NULL)) 
        vips_error_exit(NULL);
    image = t2;
    
    if (vips_image_write_to_file(image, argv[2], NULL)) 
        vips_error_exit(NULL);
            
    return 0;
}
```

I see:

```
$ gcc -g -Wall canny.c `pkg-config vips --cflags --libs`
$ ./a.out nina.jpg y.jpg 
built against libvips 8.15.3
linked to libvips 8.15.3
```

And it makes:

![y](https://github.com/user-attachments/assets/640267ac-9f9a-4040-acc2-bf449eb8e3fa)

@lovell is probably correct, I think you must be picking up a very old libvips at runtime.

**Comment by Jacherr** [comment]
`vips_version_string()` *does* report 8.15.2, weirdly. I have no idea what could have possibly happened with the installation. I suppose I'll try pruning all of the vips files I can find and reinstalling to see if that helps.

**Comment by Jacherr** [comment]
So I did a bit more digging, and it seems to be a symptom caused by the fact I am accessing vips across multiple threads. Essentially, I am bulk edge detecting many images at once using `canny`, and most of them seem to have this error (except one, curiously enough). If I change the application to be singlethreaded then this error goes away. As I understand it, vips should be thread safe as long as `vips_init()` is called in that particular thread? I did have a cursory look but didn't seem to find much in this regard?

**Comment by Jacherr** [comment]
Ah, found the cause. I wasn't calling `vips_init()` on program startup. Apologies for the confusion 😅

**Reply by jcupitt** [reply]
Ah! Makes sense. Yes, some thread stuff is initialised on `vips_init()` and locking can break if you don't call it. Glad you got it working.
